snuba/utils/streams/batching.py

```diff
@@ -61,6 +62,150 @@ class Offsets:
     hi: int  # exclusive
 
 
+@dataclass
+class Batch(Generic[TResult]):
+    results: MutableSequence[TResult] = field(default_factory=list)
+    offsets: MutableMapping[Partition, Offsets] = field(default_factory=dict)
+    created: float = field(default_factory=lambda: time.time())
+    messages_processed_count: int = 0
+    # the total amount of time, in milliseconds, that it took to process
+    # the messages in this batch (does not included time spent waiting for
+    # new messages)
+    processing_time_ms: float = 0.0
+
+
+class BatchProcessor:
+    """
+    The ``BatchProcessor`` is a message processor that accumulates processed
+    message values, periodically flushing them after a given duration of time
+    has passed or number of output values have been accumulated.
+    """
+
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        worker: AbstractBatchWorker[TPayload, TResult],
+        max_batch_size: int,
+        max_batch_time: int,
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__consumer = consumer
+        self.__worker = worker
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+        self.__metrics = metrics
+
+        self.__batch: Optional[Batch[TResult]] = None
+        self.__closed = False
+
+    def process(self, message: Optional[Message[TPayload]]) -> None:
+        """
+        Process a message or heartbeat.
+
+        Calling this method may trigger an in-flight batch to be flushed.
+        """
+        assert not self.__closed
+
+        # If we have an active batch, check if it's ready to be flushed.
+        if self.__batch is not None and (
+            len(self.__batch.results) >= self.__max_batch_size
+            or time.time() > self.__batch.created + self.__max_batch_time / 1000.0
+        ):
+            self.__flush()
+
+        # If this was just a heartbeat, there is nothing to process and we can
+        # skip the rest of the method.
+        if message is None:
+            return
+
+        start = time.time()
+
+        self.__metrics.timing(
+            "receive_latency",
+            (start - message.timestamp.timestamp()) * 1000,
+            tags={
+                "topic": message.partition.topic.name,
+                "partition": str(message.partition.index),
+            },
+        )
+
+        # Create the batch only after the first message is seen.
+        if self.__batch is None:
+            self.__batch = Batch()
```

Comment:
Since self.__worker.process_message(message) can return None, is that any reason for creating the batch before actually processing the message and being sure we have something to add to the batch? Or is there ay issue in potentially flushing an empty batch ?